product-configuration-extension/src/ProductDetailsConfigurationExtension.liquid

```diff
@@ -2,38 +2,186 @@
 import {
   reactExtension,
   useApi,
+  BlockStack,
   Text,
 } from '@shopify/ui-extensions-react/admin';
+import {useState, useEffect} from 'react';
 
 // The target used here must match the target used in the extension's toml file (./shopify.extension.toml)
-{% if flavor contains "typescript" %}
-export default reactExtension<any>('admin.product-details.configuration.render', () => <App />);
-{% else %}
 export default reactExtension('admin.product-details.configuration.render', () => <App />);
-{% endif %}
+
 function App() {
   {% if flavor contains "typescript" %}
   const {extension: {target}, i18n} = useApi<'admin.product-details.configuration.render'>();
   {% else %}
   const {extension: {target}, i18n} = useApi();
   {% endif %}
+  const product = useProduct();
   return (
-    <Text>
-      {i18n.translate('welcome', {target})}
-    </Text>
+    <BlockStack>
+      <Text>
+        {i18n.translate('welcome', {target})}
+      </Text>
+      {product?.bundleComponents.map((component) =>
+        <Text key={component.id}>{component.title}</Text>
+      )}
+    </BlockStack>
   );
 }
 
+function useProduct() {
+  {% if flavor contains "typescript" %}
+  const {data, query} = useApi<'admin.product-details.configuration.render'>();
+  const productId = (data as any)?.selected[0].id;
+  const [product, setProduct] = useState<{
+    id: string;
+    title: string;
+    bundleComponents: {
+      id: string;
+      title: string;
+    }[];
+  }>(null);
+  {% else %}
+  const {data, query} = useApi();
+  const productId = data?.selected[0].id;
+  const [product, setProduct] = useState(null);
+  {% endif %}
+
+  useEffect(() => {
+    query(
+      `#graphql
+      query GetProduct($id: ID!) {
+        product(id: $id) {
+          id
+          title
+          bundleComponents(first: 100) {
+            nodes {
+              componentProduct {
+                id
+                title
+              }
+            }
+          }
+        }
+      }
+      `,
+      {variables: {id: productId}}
+    ).then(({data, errors}) => {
+      if (errors) {
+        console.error(errors);
+      } else {
+        {% if flavor contains "typescript" %}
+        const {bundleComponents, ...product} = (data as any).product;
+        {% else %}
+        const {bundleComponents, ...product} = data.product;
+        {% endif %}
+        setProduct({
+          ...product,
+          {% if flavor contains "typescript" %}
+          bundleComponents: bundleComponents.nodes.map(({componentProduct}: any) => ({
+          {% else %}
+          bundleComponents: bundleComponents.nodes.map(({componentProduct}) => ({
+          {% endif %}
+            ...componentProduct
+          }))
+        })
+      }
+    })
+  }, [productId]);
```

Comment:
Should `query` be included in the dependencies?

Reply:
oops, yeah, it probably should be